woocommerce/woocommerce-ios
# Refunded Products: summarize algorithm

Part 1 of #1717.

So the first step is to get it working. I can successfully smush the data together to create aggregate product cells. It's now readable enough to start making changes 😀 

cc: @koke @jleandroperez (if can spare the time)

### The problem in pictures

The data provided by the API makes the Refunded Products table render like this:
<img src="https://user-images.githubusercontent.com/1062444/72114558-9eca4600-3309-11ea-9512-49853fa3c409.png" width="350" />

The aggregate data, which is what the user wants to see, should render like this:
<img src="https://user-images.githubusercontent.com/1062444/72114605-c7ead680-3309-11ea-9be6-e86dec1eaaec.png" width="350" />

### Changes
- I added a new class, `OrderItemRefundSummary`. It's based off of the struct `OrderItemRefund`.
- The `OrderItemRefundSummaryViewModel` is used for configuring the `ProductDetailsTableViewCell`.
- [summedItems](https://github.com/woocommerce/woocommerce-ios/pull/1720/files#diff-ef4a45e82a4b0b7d3eae5834c6f6d07cR18) is where the calculations happen. 

### To Test

The goal is to create multiple refunds with several of the same product type(s) in the refunds.

1. On your test site, have an order with multiple product types. Make sure to note the Order Number.
2. Find the order and "refund by item".
3. Make another refund. Refund a couple more of the same product, and add another product to make it interesting.
4. Make another refund. Refund a couple more of the same product(s) or refund a different item by itself.
5. Check out the develop branch and navigate in the app to: Orders tab > Order Number (Order Details screen). Scroll down a bit until you see the "N Items" cell. [It looks like this](https://user-images.githubusercontent.com/1062444/72114781-45aee200-330a-11ea-8774-cd437ae49d6a.png). Tap the cell to navigate to the Refunded Products screen. Notice that there are multiple cells.
6. Check out this branch. Navigate to the same location. The cells should now be summarized.

### Concerns
I'm still learning about Big O notation. If I understand it right, what I have is approaching (has arrived at?) a quadratic time complexity. Considering I'll have to re-use this algorithm in the Order Details page, that's not good.

Update release notes:
